ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/bounded/BoundedTimedQueue.java

```diff
@@ -0,0 +1,163 @@
+        this.metricsSystem.createCounter(
+            PantheonMetricCategory.EXECUTORS,
+            metricName + "_rejected_tasks_total",
+            "Total number of tasks rejected by this working queue.");
+  }
+
+  @Override
+  public boolean offer(final Runnable runnable) {
+    try {
+      requireNonNull(runnable);
+      if (!(runnable instanceof TimedTask)) {
+        return false;
+      }
+      final TimedTask task = (TimedTask) runnable;
+      if (!isEligible(task)) {
+        LOG.debug("Task does not match time policy, rejecting task.");
+        return false;
+      }
+      final boolean offered = super.offer(task);
+      if (offered) {
+        return true;
+      }
+      // Try to evict tasks based on time policy.
+      int evictedTaskCounter = processEviction();
+      if (evictedTaskCounter == 0) {
+        // No task has been evicted, rejecting new task.
+        LOG.warn(
+            "Task has been rejected because of maximum capacity of the queue has been reached and no candidate task to remove from the working queue.");
+        return false;
+      }
+      LOG.warn("{} task(s) evicted from the working queue because too old.", evictedTaskCounter);
+      totalEvictedTaskCounter.inc(evictedTaskCounter);
+      // Check if task is still eligible after eviction process.
+      if (isEligible(task) && super.offer(task)) {
+        return true;
+      } else {
+        LOG.warn("Task insertion failed after eviction process, rejecting the task.");
+        return false;
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("Eviction aborted.", e);
+      return false;
+    }
+  }
+
+  /**
+   * Process eviction of tasks too old given keepAlive.
+   *
+   * @return The number of tasks evicted.
+   * @throws InterruptedException
+   */
+  private int processEviction() throws InterruptedException {
+    int evictedTaskCounter = 0;
+    boolean continueEviction = true;
+    TimedTask lastPolledTask = null;
+    // Removing tasks that are too old.
+    while (continueEviction) {
+      lastPolledTask = (TimedTask) poll(evictionPollTimeout, evictionPollTimeoutUnit);
+      if (lastPolledTask == null) {
+        continueEviction = false;
+      } else {
+        continueEviction = isTooOld(lastPolledTask);
+        if (continueEviction) {
+          evictedTaskCounter++;
+        }
+      }
+    }
+    // Re inject last polled task if not too old.
+    if (lastPolledTask != null && !isTooOld(lastPolledTask)) {
+      super.addFirst(lastPolledTask);
+    }
+    return evictedTaskCounter;
+  }
+
+  /**
+   * Check if the given task is eligible for insertion in the queue.
+   *
+   * @param task The task to check.
+   * @return If the task is eligible.
+   */
+  private boolean isEligible(final TimedTask task) {
+    return !isTooOld(task);
+  }
+
+  private boolean isTooOld(final TimedTask task) {
+    return task.age().get(keepAliveUnit) >= keepAlive;
+  }
+
+  public static final class Default {
```

Comment:
This internal class is redundant - typically in Pantheon we'd just make these public static fields on the `BoundedQueue` class.

Reply:
Done.